### PR TITLE
Use early-serialising KV implementation by default

### PR DESCRIPTION
**Pull request by eddyashton**
Following #1216, this:

- De-templates `kv::Map` (and `kv::TxView`), making them operate _only_ on `byte vector -> byte vector`
- Renames `kv::Map` (and `kv::TxView`) to `kv::untyped::` (and the files renamed to `kv/untyped_`)
- Renames `kv::experimental::Map` to `kv::Map` (and deletes now unused `experimental.h`)
- Makes all the governance tables use Json to write their Ks and Vs, rather than MsgPack

The resulting semantic change is that _all_ of our maps do early-serialisation/late-deserialisation (at the point of `put`/`get`, not at communication with the ledger). Additionally, our governance tables are more easily readable in the ledger, since they're stored as Json (this also means they're a bit bigger, with byte-vectors particularly enlarged - we should look at storing these as text where appropriate).

The resulting diff is awkward. Its _mostly_ copying and pasting between files, but complicated by some additional shuffling, and because `kv/map.h` has both been swapped out-of and swapped in-to. I'll try to clarify the changes inline.

### CODE REVIEW

File: src/apps/js_generic/js_generic.cpp [hunk 1]
```diff
@@ -13,7 +14,7 @@ namespace ccfapp
   using namespace kv;
   using namespace ccf;
 
-  using Table = kv::Map<std::vector<uint8_t>, std::vector<uint8_t>>;
+  using Table = kv::untyped::Map;
```
Comment (eddyashton) [inline src/apps/js_generic/js_generic.cpp:17]: This is the only instance of this, but its something we probably want: the JS app doesn't need typed wrappers at the C++ level, its tables are bytes-to-bytes. Serialisation can be handled within the JS code.

File: src/kv/change_set.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (eddyashton) [inline src/kv/change_set.h:1]: This was the top of `tx_view.h`, now it has its own file.

File: src/kv/kv_types.h [hunk 3]
```diff
@@ -31,6 +36,8 @@ namespace kv
   using BatchVector = std::vector<
     std::tuple<kv::Version, std::shared_ptr<std::vector<uint8_t>>, bool>>;
 
+  using SerialisedEntry = std::vector<uint8_t>;
```
Comment (eddyashton) [inline src/kv/kv_types.h:39]: This was `kv::experimental::SerialisedRep`. I decided `SerialisedEntry` was a clearer name, and it gets promoted to `kv::` as `experimental::` dies.

File: src/kv/store.h [hunk 4]
```diff
@@ -147,12 +147,12 @@ namespace kv
...
      */
-    template <class K, class V, class H = std::hash<K>>
-    Map<K, V, H>& create(
+    template <class K, class V>
+    Map<K, V>& create(
       std::string name,
       SecurityDomain security_domain = kv::SecurityDomain::PRIVATE)
     {
-      return create<Map<K, V, H>>(name, security_domain);
+      return create<Map<K, V>>(name, security_domain);
```
Comment (eddyashton) [inline src/kv/store.h:155]: Although this looks almost identical, it is creating a new type of `Map`. The third argument to `kv::Map` is now `KSerialiser`, not `Hasher`, so `H` is removed here.

File: src/kv/test/kv_test.cpp [hunk 5]
```diff
@@ -15,73 +14,62 @@
...
-struct ExperimentalMapTypes
-{
-  using StringString = kv::experimental::Map<std::string, std::string>;
-  using NumNum = kv::experimental::Map<size_t, size_t>;
-  using NumString = kv::experimental::Map<size_t, std::string>;
-  using StringNum = kv::experimental::Map<std::string, size_t>;
-};
-
-TEST_CASE_TEMPLATE("Map creation", MapImpl, RawMapTypes, ExperimentalMapTypes)
+TEST_CASE("Map creation")
```
Comment (eddyashton) [inline src/kv/test/kv_test.cpp:25]: This file is just a regex replace: `TEST_CASE_TEMPLATE(XXX, ...)` becomes `TEST_CASE(XXX)`.

File: src/kv/tx_view.h [hunk 6]
```diff
@@ -2,266 +2,83 @@
...
-      committed(committed_state),
-      start_version(current_version)
-    {}
-
-    ChangeSet(ChangeSet&) = delete;
-  };
-
-  template <typename K, typename V, typename H = std::hash<K>>
-  class TxView
+  template <typename K, typename V, typename KSerialiser, typename VSerialiser>
```
Comment (eddyashton) [inline src/kv/tx_view.h:11]: Useless diff. The left side is now `src/kv/untyped_tx_view.h`, minus the definitions that were pulled into `src/kv/change_set.h`. The right side was in `experimental.h`.

File: src/kv/untyped_tx_view.h [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+  using VersionV = kv::VersionV<SerialisedEntry>;
+  using State =
+    kv::State<SerialisedEntry, SerialisedEntry, SerialisedKeyHasher>;
+  using Read = kv::Read<SerialisedEntry, SerialisedKeyHasher>;
+  using Write =
+    kv::Write<SerialisedEntry, SerialisedEntry, SerialisedKeyHasher>;
+  using ChangeSet =
+    kv::ChangeSet<SerialisedEntry, SerialisedEntry, SerialisedKeyHasher>;
+
+  class TxView
```
Comment (eddyashton) [inline src/kv/untyped_tx_view.h:22]: This was in `tx_view.h`, and now its here with no template.

File: src/node/certs.h [hunk 8]
```diff
@@ -5,5 +5,5 @@
 
 namespace ccf
 {
-  using Certs = kv::Map<Cert, ObjectId>;
+  using Certs = kv::JsonSerialisedMap<Cert, ObjectId>;
```
Comment (eddyashton) [inline src/node/certs.h:8]: This is what all the changes look like for our governance tables. If they were `kv::Map`, they'd write their Ks and Vs as msgpack, but we want Json so the type of the Map (and only the Map) changes.

We could now go through and remove a bunch of `MSGPACK_DEFINE`s - these types only need to be Json-able, not MsgPack-able. But maybe we'll change our minds, so the `MSGPACK` transformations can stay for now.

File: src/kv/untyped_map.h [hunk 9]
```diff
@@ -0,0 +1,734 @@
...
+  };
+  using LocalCommits = snmalloc::DLList<LocalCommit, std::nullptr_t, true>;
+
+  struct Roll
+  {
+    std::unique_ptr<LocalCommits> commits;
+    size_t rollback_counter;
+  };
+
+  struct RollContainer
```
Comment (eddyashton) [inline src/kv/untyped_map.h:79]: Template dependent lookup let us get away with some circular dependencies between `kv::Map<K, V>` and `kv::TxViewCommitter<K, V>`. This doesn't work now they're non-templated, `kv::Map` and `kv::untyped::TxView>`. To break the loop we create this little interface, which is how a `TxViewCommitter` sees its parent `Map`. No more friendship, just a vague `RollContainer`.

File: src/kv/untyped_map.h [hunk 10]
```diff
@@ -0,0 +1,734 @@
...
+      Version v) :
+      TxViewCommitter(m, rollbacks, current_state, committed_state, v),
+      TxView(TxViewCommitter::change_set)
+    {}
+  };
+
+  class Map : public AbstractMap, public RollContainer
+  {
+  public:
+    using K = SerialisedEntry;
```
Comment (eddyashton) [inline src/kv/untyped_map.h:264]: Although the Key and Value are the same type, I've retained the K and V names so its clear which we _intend_ to use within this class.

File: src/kv/test/kv_serialisation.cpp [hunk 11]
```diff
@@ -236,112 +212,19 @@ TEST_CASE_TEMPLATE(
 }
 
 struct CustomClass
-{
```
Comment (eddyashton) [inline src/kv/test/kv_serialisation.cpp:239]: This file is mostly a find and replace, plus deletions like this for things that only tested the old implementation.

File: src/kv/map.h [hunk 12]
```diff
@@ -2,723 +2,202 @@
...
-    }
-  }
-
-  template <class K, class V, class H = std::hash<K>>
-  class Map;
-
-  template <class K, class V, class H>
-  class TxViewCommitter : public AbstractTxView
+  template <typename K, typename V, typename KSerialiser, typename VSerialiser>
+  class TypedMap : public AbstractMap
```
Comment (eddyashton) [inline src/kv/map.h:13]: This was `kv::experimental::Map`, now its `kv::TypedMap`, providing the `kv::Map` replacement via the typedefs at the bottom of this file.

File: src/kv/map.h [hunk 13]
```diff
@@ -2,723 +2,202 @@
...
+  using MsgPackSerialisedMap = TypedMap<
+    K,
+    V,
+    kv::serialisers::MsgPackSerialiser<K>,
+    kv::serialisers::MsgPackSerialiser<V>>;
+
+  // The default kv::Map will use msgpack serialisers. Custom types are
+  // supported through the MSGPACK_DEFINE macro
+  template <typename K, typename V>
+  using Map = MsgPackSerialisedMap<K, V>;
```
Comment (eddyashton) [inline src/kv/map.h:202]: Specifying serialisers is slightly verbose, so we name some sensible defaults for you. And finally we provide `kv::Map<K, V>` again here. Note that this is a slight breaking change - anyone who had a custom hasher (ie `kv::Map<K, V, H>`) will get a compile error, and they'll have to remove it. We had no instances of this, and user custom types should no longer need it.

File: src/apps/js_generic/js_generic.cpp (hunk 1 above)
Comment (achamayou) [reply src/apps/js_generic/js_generic.cpp:17]: Yes, that's very much the intention!

File: src/node/certs.h (hunk 8 above)
Comment (achamayou) [reply src/node/certs.h:8]: Let's have a look at the cost of decoding the Cert from JSON on every lookup.

**Comment by eddyashton** [comment]
@achamayou Agreed, this performance hit is unexpected. The certs may account for the e2e test, but clearly something's much slower in the standalone benchmark as well. Converting to Draft while I investigate this.

**Comment by achamayou** [comment]
@eddyashton yeah no, it's not just the certs, there's something else for sure.

File: src/kv/serialise_entry_json.h [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+#include <nlohmann/json.hpp>
+#include <vector>
+
+namespace kv::serialisers
+{
+  template <typename T>
+  struct JsonSerialiser
+  {
+    using Bytes = std::vector<uint8_t>;
```
Comment (achamayou) [inline src/kv/serialise_entry_json.h:13]: If we're going to have Bytes, we should probably have it somewhere common to at least the serialisers.

**Comment by eddyashton** [comment]
Closing this PR and re-opening a clean replacement.

**Pull request closed without merging**